File: src/tachikoma/projects/tools.py
"""MCP tools for project management.

Provides tools for registering and deregistering project submodules
during conversations.
"""

import contextlib
from pathlib import Path

from claude_agent_sdk import create_sdk_mcp_server, tool
from claude_agent_sdk.types import McpSdkServerConfig
from loguru import logger

from tachikoma.projects.git import (
    add_submodule,
    checkout_branch,
    has_uncommitted_changes_detail,
    remove_submodule,
    resolve_default_branch,
)

_log = logger.bind(component="projects")


def create_projects_server(workspace_path: Path) -> McpSdkServerConfig:
    """Create SDK MCP server with project management tools.

    The tools have closure over the workspace_path for git operations.

    Args:
        workspace_path: The workspace root directory.

    Returns:
        McpSdkServerConfig for use with ClaudeAgentOptions.mcp_servers.
    """

    @tool(
        "register_project",
        "Register a new project as a git submodule under projects/<name>.",
        {"name": str, "url": str},
    )
    async def register_project(args: dict) -> dict:
        """Register a new project by adding it as a git submodule.

        Args:
            args: Must contain 'name' (project name) and 'url' (git remote URL).

        Returns:
            Success message or error with is_error=True.
        """
        name = args.get("name", "")
        url = args.get("url", "")

        if not name:
            return {
                "content": [{"type": "text", "text": "Error: 'name' is required"}],
                "is_error": True,
            }

        if not url:
            return {
                "content": [{"type": "text", "text": "Error: 'url' is required"}],
                "is_error": True,
            }

        project_path = workspace_path / "projects" / name

        if project_path.exists():
            return {
                "content": [
                    {"type": "text", "text": f"Error: Project '{name}' already exists"}
                ],
                "is_error": True,
            }

        try:
            await add_submodule(workspace_path, name, url)
            default_branch = await resolve_default_branch(project_path)
            await checkout_branch(project_path, default_branch)

            _log.info(
                "Project registered: name={name} url={url} branch={branch}",
                name=name,
                url=url,
                branch=default_branch,
            )

            return {
                "content": [
                    {
                        "type": "text",
                        "text": f"Registered project '{name}' (branch: {default_branch}). "
                        f"The project is now available under projects/{name}. "
                        f"Changes will be committed to the workspace at session end.",
                    }
                ]
            }

        except Exception as e:
            # Cleanup partial state on failure
            if project_path.exists():
                with contextlib.suppress(Exception):
                    await remove_submodule(workspace_path, name)

            error_msg = str(e)
            _log.warning(
                "Project registration failed: name={name} err={err}",
                name=name,
                err=error_msg,
            )

            return {
                "content": [
                    {"type": "text", "text": f"Error registering project: {error_msg}"}
                ],
                "is_error": True,
            }

    @tool(
        "deregister_project",
        "Remove a registered project. Warns if uncommitted changes exist unless force=true.",
        {"name": str, "force": bool},
    )
    async def deregister_project(args: dict) -> dict:
        """Deregister a project by removing the git submodule.

        Args:
            args: Must contain 'name'. Optional 'force' (default false).

        Returns:
            Success message, warning, or error with is_error=True.
        """
        name = args.get("name", "")
        force = args.get("force", False)

        if not name:
            return {
                "content": [{"type": "text", "text": "Error: 'name' is required"}],
                "is_error": True,
            }

        project_path = workspace_path / "projects" / name

        if not project_path.exists():
            return {
                "content": [
                    {"type": "text", "text": f"Error: Project '{name}' not found"}
                ],
                "is_error": True,
            }

        try:
            changes = await has_uncommitted_changes_detail(project_path)

            if changes and not force:
                return {
                    "content": [
                        {
                            "type": "text",
                            "text": f"Warning: Project '{name}' has uncommitted changes:\n"
                            f"{changes}\n\n"
                            f"Use force=true to remove anyway (changes will be lost).",
                        }
                    ],
                    "is_error": True,
                }

            await remove_submodule(workspace_path, name)

            _log.info("Project deregistered: name={name} force={force}", name=name, force=force)

            return {
                "content": [
                    {
                        "type": "text",
                        "text": f"Deregistered project '{name}'. "
                        f"Changes will be committed to the workspace at session end.",
                    }
                ]
            }

        except Exception as e:
            error_msg = str(e)
            _log.warning(
                "Project deregistration failed: name={name} err={err}",
                name=name,
                err=error_msg,
            )

            return {
                "content": [
                    {"type": "text", "text": f"Error deregistering project: {error_msg}"}
                ],
                "is_error": True,
            }

    return create_sdk_mcp_server(
        name="projects",
        version="1.0.0",
        tools=[register_project, deregister_project],
    )

File: src/extensions/projects/tools.ts
import { access } from "node:fs/promises";
import { join } from "node:path";

import type { ExtensionFactory } from "@earendil-works/pi-coding-agent";
import { type Static, Type } from "typebox";

import type { Logger } from "../../log.ts";
import {
  addSubmodule,
  checkoutBranch,
  describeProjectState,
  listSubmodules,
  projectState,
  removeSubmodule,
  resolveDefaultBranch,
  uncommittedChangesDetail,
} from "./git.ts";

export interface ProjectToolDeps {
  workspaceRoot: string;
  log: Logger;
}

export const RegisterProjectParams = Type.Object({
  name: Type.String({ description: "Project name — becomes the directory under projects/" }),
  url: Type.String({ description: "Git remote URL to clone the project from" }),
});

export const DeregisterProjectParams = Type.Object({
  name: Type.String({ description: "Name of the registered project to remove" }),
  force: Type.Optional(
    Type.Boolean({
      description: "Remove even when the project has uncommitted changes (they will be lost)",
    }),
  ),
});

export const ListProjectsParams = Type.Object({});

const exists = async (path: string): Promise<boolean> =>
  access(path).then(
    () => true,
    () => false,
  );

export const handleRegisterProject = async (
  { workspaceRoot, log }: ProjectToolDeps,
  args: Static<typeof RegisterProjectParams>,
): Promise<string> => {
  if (args.name === "") throw new Error("'name' is required");
  if (args.url === "") throw new Error("'url' is required");

  const projectPath = join(workspaceRoot, "projects", args.name);

  if (await exists(projectPath)) throw new Error(`Project '${args.name}' already exists`);

  try {
    await addSubmodule(workspaceRoot, args.name, args.url);

    const defaultBranch = await resolveDefaultBranch(projectPath);
    await checkoutBranch(projectPath, defaultBranch);

    log.info({ name: args.name, url: args.url, branch: defaultBranch }, "project registered");

    return (
      `Registered project '${args.name}' (branch: ${defaultBranch}). ` +
      `The project is now available under projects/${args.name}. ` +
      "Changes will be committed to the workspace at session end."
    );
  } catch (error) {
    if (await exists(projectPath)) {
      try {
        await removeSubmodule(workspaceRoot, args.name);
      } catch {
        // Partial-state cleanup is best-effort; the original error matters more.
      }
    }

    log.warn({ name: args.name, err: error }, "project registration failed");
    throw new Error(`Error registering project: ${(error as Error).message}`);
  }
};

export const handleDeregisterProject = async (
  { workspaceRoot, log }: ProjectToolDeps,
  args: Static<typeof DeregisterProjectParams>,
): Promise<string> => {
  if (args.name === "") throw new Error("'name' is required");

  const projectPath = join(workspaceRoot, "projects", args.name);

  if (!(await exists(projectPath))) throw new Error(`Project '${args.name}' not found`);

  const force = args.force ?? false;
  const changes = await uncommittedChangesDetail(projectPath);

  if (changes != null && !force) {
    throw new Error(
      `Project '${args.name}' has uncommitted changes:\n${changes}\n\n` +
        "Use force=true to remove anyway (changes will be lost).",
    );
  }

  await removeSubmodule(workspaceRoot, args.name);

  log.info({ name: args.name, force }, "project deregistered");

  return (
    `Deregistered project '${args.name}'. ` +
    "Changes will be committed to the workspace at session end."
  );
};

export const handleListProjects = async ({ workspaceRoot }: ProjectToolDeps): Promise<string> => {
  const submodulePaths = await listSubmodules(workspaceRoot);

  if (submodulePaths.length === 0) {
    return "No projects registered. Use register_project to add one.";
  }

  const lines = await Promise.all(
    submodulePaths.map(async (path) =>
      describeProjectState(await projectState(workspaceRoot, path)),
    ),
  );

  return `# Registered Projects\n\n${lines.join("\n")}`;
};

const textResult = (text: string) => ({
  content: [{ type: "text" as const, text }],
  details: undefined,
});

/** pi extension factory exposing the project management tools to the agent. */
export const createProjectsToolsFactory =
  (deps: ProjectToolDeps): ExtensionFactory =>
  (pi) => {
    pi.registerTool({
      name: "register_project",
      label: "Register Project",
      description:
        "Register an external git repository as a project: clones it as a submodule under projects/<name> and checks out its default branch. Project changes are committed and pushed automatically at session end.",
      promptSnippet: "Register an external git repository as a tracked project",
      promptGuidelines: [
        "Use register_project when the user wants to work on an external repository inside the workspace.",
      ],
      parameters: RegisterProjectParams,
      async execute(_toolCallId, params) {
        return textResult(await handleRegisterProject(deps, params));
      },
    });

    pi.registerTool({
      name: "deregister_project",
      label: "Deregister Project",
      description:
        "Remove a registered project from the workspace. Fails when the project has uncommitted changes unless force=true is passed (those changes are lost).",
      promptSnippet: "Remove a registered project from the workspace",
      promptGuidelines: [
        "Confirm with the user before calling deregister_project with force=true — uncommitted work is lost.",
      ],
      parameters: DeregisterProjectParams,
      async execute(_toolCallId, params) {
        return textResult(await handleDeregisterProject(deps, params));
      },
    });

    pi.registerTool({
      name: "list_projects",
      label: "List Projects",
      description:
        "List the registered projects with their current branch (or detached commit) and uncommitted-change counts.",
      promptSnippet: "List registered projects and their git state",
      promptGuidelines: [
        "Check list_projects before register_project to avoid registering a duplicate.",
      ],
      parameters: ListProjectsParams,
      async execute() {
        return textResult(await handleListProjects(deps));
      },
    });
  };
